### ISSUE TITLE
Google Places address autocomplete javascript initialisation bug

### ISSUE THREAD

**Issue by simple-nicholas** [verbb/formie#1109]
### Describe the bug

Address autocomplete javascript _very_ inconsistently throws an error (`Uncaught TypeError: Cannot read properties of undefined (reading 'places')`) in initialisation, breaking address autocomplete for the whole form.

The error is in the incorrect use of onload in the initScript function called by the constructor of class FormieGoogleAddress (google-address.js line 41).

The script elements src `https://maps.googleapis.com/maps/api/js?key=_______&libraries=places` is a bootstrap which loads the core and extra libraries requested. So the onload which calls the initAutocomplete function can trigger any time after the bootstrap is loaded, including before the Google libraries are ready to use. But initAutocomplete requires these libraries to be available (Line 70), causing a race condition.

### Steps to reproduce

1. Create a form with an address field using autocomplete powered by Google Places
2. Open the form in a browser
3. Multiple reloads may be required to observe the console error `Cannot read properties of undefined (reading 'places')`

This issue is very inconsistent to reproduce, Incognito mode and disabling caching seems to work best.

### Form settings

- Multi-page form: Yes
- Submission Method: Either
- Client-side Validation: Yes
- Custom Form Templates: No

### Craft CMS version

4.2.3

### Plugin version

2.0.12

### Multi-site?

No

**Comment by engram-design** [comment]
> The error is in the incorrect use of onload in the initScript function called by the constructor of class FormieGoogleAddress (google-address.js line 41).

What would be the considered correct use? The `onload` of this script should be called when the loaded scripts are finished loading, and these libraries are initialised immediately.

I can't replicate your issue, no matter what, even when throttling my internet speed or disabling cache/incognito, etc. As you say, it's inconsistent, which makes it tricky to troubleshoot. However, I have altered our scripts to wait until the Google libraries are available before initialising (we previously already did this when loading the Google scripts externally), so keen to see if that helps your issue.

To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.12",
  "...": "..."
}
```

Then run `composer update`.

**Comment by simple-nicholas** [comment]
LIke I said, it's a bootstrap which triggers the async loading of the actual libraries (you can see them at or near the bottom of the header), so onload doesn't provide what's needed here. The [Dynamic Loading](https://developers.google.com/maps/documentation/javascript/overview#js_api_loader_package) from the docs would be the simplest solution in this case.

The noted fix on the dev branch still errors sometimes as it's only testing for the base library, not Places as well.

**Comment by engram-design** [comment]
The only reason we don't use the dynamic loading as an NPM package is that some people are already loading the script externally for their site, and Formie can make use of that script. Otherwise, (from what I understand) we need to bundle the loader script as part of our front-end JS, which I'd rather not do.

We're essentially following the [Dynamic Loading (Manual)](https://developers.google.com/maps/documentation/javascript/overview#Dynamic) approach already?

But yes, fair point to check for [`google.maps.places`](https://github.com/verbb/formie/blob/bd21801152c71311db2a590abd09f83a0ea5e21a/src/web/assets/frontend/src/js/address-providers/google-address.js#L59)

**Comment by engram-design** [comment]
@simple-nicholas I've also just pushed another improvement for this, which should hopefully resolve that.

**Comment by engram-design** [comment]
Fixed in [2.0.13](https://github.com/verbb/formie/releases/tag/2.0.13)
